### PR TITLE
Adjacent

### PR DESCRIPTION
**Pull request by kmoham6**
Adapting adjacent_difference to C++20

### CODE REVIEW

**Review by hkaiser — COMMENTED** [review]
Could you please update the docs as well: 
- https://github.com/STEllAR-GROUP/hpx/blob/master/docs/sphinx/api/public_api.rst and 
- https://github.com/STEllAR-GROUP/hpx/blob/master/docs/sphinx/manual/writing_single_node_hpx_applications.rst?

**Review by hkaiser — CHANGES_REQUESTED** [review]

**Comment by hkaiser** [comment]
Also, please fix the compilation problems exposed by the CI.

File: libs/core/include_local/include/hpx/local/numeric.hpp [hunk 1]
```diff
@@ -10,5 +10,8 @@
 #include <hpx/parallel/numeric.hpp>
 
 namespace hpx {
-    using hpx::parallel::adjacent_difference;
+    using hpx::parallel::exclusive_scan;
```
Comment (Jedi18) [inline libs/core/include_local/include/hpx/local/numeric.hpp:13]: Could you remove these since they have been removed on master?

**Comment by hkaiser** [comment]
@kmoham6 could you please take care of the remaining CI problems?

**Comment by kmoham6** [comment]
@hkaiser I am working with Srinivas. I am taking care of the changed files in my branch. conflict will be fixed soon.

File: libs/core/algorithms/include/hpx/parallel/algorithms/detail/adjacent_difference.hpp [hunk 2]
```diff
@@ -27,13 +28,24 @@ namespace hpx { namespace parallel { inline namespace v1 { namespace detail {
...
+            if (first == last)
+                return dest;
+
+            using value_t = typename std::iterator_traits<InIter>::value_type;
+            value_t acc = *first;
+            *dest = acc;
+            while (++first != last)
+            {
+                value_t val = *first;
+                *++dest = op(val, std::move(acc));    // std::move since C++20
```
Comment (msimberg) [inline libs/core/algorithms/include/hpx/parallel/algorithms/detail/adjacent_difference.hpp:45]:
```suggestion
                *++dest = HPX_INVOKE(op, val, std::move(acc));
```
What is the `C++20` comment saying?

File: libs/core/algorithms/tests/unit/algorithms/adjacentdifference.cpp [hunk 3]
```diff
@@ -24,48 +26,6 @@ void adjacent_difference_test()
...
-    test_adjacent_difference_bad_alloc_async(seq(task), IteratorTag());
-    test_adjacent_difference_bad_alloc_async(par(task), IteratorTag());
-}
-
-void adjacent_difference_bad_alloc_test()
-{
-    test_adjacent_difference_bad_alloc<std::random_access_iterator_tag>();
-    test_adjacent_difference_bad_alloc<std::forward_iterator_tag>();
-}
-
```
Comment (msimberg) [inline libs/core/algorithms/tests/unit/algorithms/adjacentdifference.cpp:68]: Have these moved somewhere else or are they no longer relevant?

File: libs/core/include_local/include/hpx/local/numeric.hpp [hunk 4]
```diff
@@ -10,5 +10,5 @@
 #include <hpx/parallel/numeric.hpp>
 
 namespace hpx {
-    using hpx::parallel::adjacent_difference;
+
 }    // namespace hpx
```
Comment (msimberg) [inline libs/core/include_local/include/hpx/local/numeric.hpp:14]: You can remove the whole `namespace hpx {}` here.

File: libs/core/algorithms/tests/unit/algorithms/adjacentdifference.cpp (hunk 3 above)
Comment (kmoham6) [reply libs/core/algorithms/tests/unit/algorithms/adjacentdifference.cpp:68]: Yes, they have moved to the files below: 
libs/core/algorithms/tests/unit/algorithms/adjacentdifference_exception.cpp
libs/core/algorithms/tests/unit/algorithms/adjacentdifference_bad_alloc.cpp

**Comment by hkaiser** [comment]
@kmoham6 the cmake_format issue should be gone if you rebase on master. The other test issues are unrelated and known, you will not have to worry about those.

**Comment by msimberg** [comment]
@kmoham6 could you mark comments that you've already taken care of as resolved so that we can see what is actually still unresolved?

File: libs/core/include_local/include/hpx/local/numeric.hpp (hunk 4 above)
Comment (kmoham6) [reply libs/core/include_local/include/hpx/local/numeric.hpp]: I removed whole "namespace hpx {}"

File: libs/core/algorithms/include/hpx/parallel/algorithms/detail/adjacent_difference.hpp (hunk 2 above)
Comment (kmoham6) [reply libs/core/algorithms/include/hpx/parallel/algorithms/detail/adjacent_difference.hpp]: I removed this C++20 comment

File: libs/core/algorithms/include/hpx/parallel/algorithms/adjacent_difference.hpp [hunk 5]
```diff
@@ -16,7 +154,9 @@
 #include <hpx/executors/execution_policy.hpp>
 #include <hpx/parallel/algorithms/detail/adjacent_difference.hpp>
```
[the remark this answers is not in the document]
Comment (kmoham6) [reply libs/core/algorithms/include/hpx/parallel/algorithms/adjacent_difference.hpp:155]: I changed the name to adjacent_difference.hpp

**Comment by hkaiser** [comment]
@kmoham6 not sure if I missed it before, but now we see compilation problems with the datapar tests: https://cdash.cscs.ch/viewBuildError.php?buildid=193713

**Comment by kmoham6** [comment]
@hkaiser seems the error is about std::minus<> that I changed it before my last rebase. it looks like they were lost after rebase. I fixed them and pushing again.

**Comment by kmoham6** [comment]
> @kmoham6 not sure if I missed it before, but now we see compilation problems with the datapar tests: https://cdash.cscs.ch/viewBuildError.php?buildid=193713

**Review by hkaiser — APPROVED** [review]
This is now ready to go in. LGTM, thanks!

**Pull request merged**